src/colvar_neuralnetworkcompute.cpp

```diff
@@ -283,27 +262,1003 @@ std::vector<std::vector<double>> neuralNetworkCompute::multiply_matrix(const std
+	{
+		for (int j = 0; j < B_l; j++)
+		{
+			C[i][j+A_l]=B[i][j];
+		}
+	}
+	return C;
+}
+//transpose the matrix
+std::vector<std::vector<double>> transpose(const std::vector<std::vector<double>> &A)
+{
+	std::vector<std::vector<double>> AT=creatematrix(A[0].size(),A.size());
+	int h=AT.size();
+	int l=AT[0].size();
+	for (int i = 0; i <h ; i++)
+	{
+		for (int j = 0; j < l; j++)
+		{
+			AT[i][j]=A[j][i];
+		}
+	}
+	return AT;
+}
+//transpose the third dimension of matrix with each omitting the first corderate
+std::vector<std::vector<std::vector<double>>> transpose(const std::vector<std::vector<std::vector<double>>>& A)
+{
+	std::vector<std::vector<std::vector<double>>> AT;
+	for(int i=0;i<A.size();i++)
+	{
+		AT.push_back(transpose(A[i]));
+	}
+	return AT;
+}
+void show_matrix(const std::vector<std::vector<double>> &A)
+{
+	int h=A.size();
+	int l=A[0].size();
+	for (int i = 0; i < h; i++)
+	{
+		for (int j = 0; j < l; j++)
+		{
+			std::cout<<A[i][j]<<"\t";
+		}
+		std::cout<<std::endl;
+	}
+}
+//reshape the matrix
+//from one dimention to two dimention
+std::vector<std::vector<double>> reshape(const std::vector<double>& A, int shape0,int shape1)
+{
+    std::vector<std::vector<double>> AT=creatematrix(shape0,shape1);
+    if(A.size()/shape0!=shape1)
+    {
+        std::cout<<("ERROR: Invalid reshape with from one dimention to"+std::to_string(shape0)+"*"+std::to_string(shape1)+"\n");
+        return AT;
+    }
+    for(int i=0;i<shape0;i++)
+    {
+        for(int j=0;j<shape1;j++)
+        {
+            AT[i][j]=A[i*shape1+j];
+        }
+    }
+    return AT;
+}
+//from two dimention to two dimention
+std::vector<std::vector<double>> reshape(const std::vector<std::vector<double>>& A, int shape0,int shape1)
+{
+    std::vector<std::vector<double>> AT=creatematrix(shape0,shape1);
+    if(A.size()*A[0].size()/shape0!=shape1)
+    {
+        std::cout<<("ERROR: Invalid reshape with from two dimention to"+std::to_string(shape0)+"*"+std::to_string(shape1)+"\n");
+        return AT;
+    }
+    for(int i=0;i<shape0;i++)
+    {
+        for(int j=0;j<shape1;j++)
+        {
+            AT[i][j]=A[(i+j)/A[0].size()][(i+j)%A[0].size()];
+        }
+    }
+    return AT;
+}
+//from two dimention to one dimention, with shape0 equals to flattened variables. 
+//Notice that this is a specific function only applied into this model, because it is reversive. 
+//that is, to concatenate each seq_len*d_model matrix. 
+std::vector<double> concat(const std::vector<std::vector<double>>& A, int shape0)
+{
+    std::vector<double> AT(shape0);
+    if(A.size()*A[0].size()!=shape0)
+    {
+        std::cout<<("ERROR: Invalid concatenate with from two dimention to"+std::to_string(shape0)+" dimention\n");
+        return AT;
+    }
+    int seq_len=A.size();
+    int d_model=A[0].size();
+
+    for(int i=0;i<d_model;i++)
+    {
+        for(int j=0;j<seq_len;j++)
+        {
+            AT[i*seq_len+j]=A[j][i];
+        }
+    }
+    return AT;
```

Comment:
This `concat` is very different from what I have used in pytorch. Could you write some more comments about how it supposes to work?